langchain_tools_demo/orchestration/langchain_tools/agent.py

```diff
@@ -14,101 +14,101 @@
 
 set_verbose(bool(os.getenv("DEBUG", default=False)))
 BASE_URL = os.getenv("BASE_URL", default="http://127.0.0.1:8080")
 
-# aiohttp context
-connector = None
-
 CLOUD_RUN_AUTHORIZATION_TOKEN = None
 
 
-# Class for setting up a dedicated llm agent for each individual user
-class UserAgent:
+class Orchestration(orchestration.Orchestration):
     client: aiohttp.ClientSession
     agent: AgentExecutor
 
-    def __init__(self, client, agent) -> None:
+    @orchestration.classproperty
+    def kind(cls):
+        return "langchain-tools"
+
+    def __init__(self, client: aiohttp.ClientSession, agent: AgentExecutor):
         self.client = client
         self.agent = agent
 
-
-user_agents: Dict[str, UserAgent] = {}
-
-
-async def get_connector():
-    global connector
-    if connector is None:
-        connector = aiohttp.TCPConnector(limit=100)
-    return connector
-
-
-async def handle_error_response(response):
-    if response.status != 200:
-        return f"Error sending {response.method} request to {str(response.url)}): {await response.text()}"
-
-
-async def create_client_session() -> aiohttp.ClientSession:
-    return aiohttp.ClientSession(
-        connector=await get_connector(),
-        connector_owner=False,
-        headers={},
-        raise_for_status=True,
-    )
-
-
-# Agent
-async def init_agent(history: list[messages.BaseMessage]) -> UserAgent:
-    """Load an agent executor with tools and LLM"""
-    print("Initializing agent..")
-    llm = VertexAI(max_output_tokens=512, model_name="gemini-pro")
-    memory = ConversationBufferMemory(
-        chat_memory=ChatMessageHistory(messages=history),
-        memory_key="chat_history",
-        input_key="input",
-        output_key="output",
-    )
-    client = await create_client_session()
-    tools = await initialize_tools(client)
-    agent = initialize_agent(
-        tools,
-        llm,
-        agent=AgentType.STRUCTURED_CHAT_ZERO_SHOT_REACT_DESCRIPTION,
-        memory=memory,
-        handle_parsing_errors=True,
-        max_iterations=3,
-        early_stopping_method="generate",
-        return_intermediate_steps=True,
-    )
-    # Create new prompt template
-    tool_strings = "\n".join([f"> {tool.name}: {tool.description}" for tool in tools])
-    tool_names = ", ".join([tool.name for tool in tools])
-    format_instructions = FORMAT_INSTRUCTIONS.format(
-        tool_names=tool_names,
-    )
-    today_date = date.today().strftime("%Y-%m-%d")
-    today = f"Today is {today_date}."
-    template = "\n\n".join([PREFIX, tool_strings, format_instructions, SUFFIX, today])
-    human_message_template = "{input}\n\n{agent_scratchpad}"
-    prompt = ChatPromptTemplate.from_messages(
-        [("system", template), ("human", human_message_template)]
-    )
-    agent.agent.llm_chain.prompt = prompt  # type: ignore
-
-    return UserAgent(client, agent)
+    @classmethod
+    async def create(cls, base_history: list[Any]) -> "Orchestration":
```

Comment:
We need to separate the "Orchestration" from the agents inside of it. The "Orchestrator" layer should be able to handle everything, including getting agents